### PR TITLE
DIP-279 Trigger File

### PR DESCRIPTION
**Pull request by cameron-white-janrain**

### CODE REVIEW

File: bluekai/bluekai_trigger.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from hashlib import md5
+
+def fromData(filename, data_raw):
+    data = bytes(data_raw, 'utf8')
+    data_size = len(data)
```
Comment (BCox-Janrain) [inline bluekai/bluekai_trigger.py]: I know `data_size` is the size of the data in bytes, but did you verify that it's also the size of the file on disk?

Comment (cameron-white-janrain) [inline bluekai/bluekai_trigger.py]: Yes I did. Here is proof.

`python3 -c "print(len(bytes('abc', 'utf8')))"`
3
`echo -n "abc" > test.txt && ls -l test.txt | awk '{ print $5 }'`
3

File: bluekai/bluekai_writter.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
+from . import bluekai_trigger
+from .config import remote_filename
+
+class BlueKaiWritter():
```
Comment (ebeesonjanrain) [inline bluekai/bluekai_writter.py]: I hate to be that guy, but it's spelled 'writer'
I keep having corn fritters popping into my brain now...

File: tests/test_bluekai_writter.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+from unittest import TestCase
+from datetime import datetime
+from mock import MagicMock
+from mock import call
+from bluekai.bluekai_writter import *
+from freezegun import freeze_time
+
```
Comment (ebeesonjanrain) [inline tests/test_bluekai_writter.py]: freezegun is pretty cool (no pun intended)

File: bluekai/bluekai_writter.py (hunk 2 above)
Comment (BCox-Janrain) [inline bluekai/bluekai_writter.py]: I was planning on making a commit for that after checking the code. :stuck_out_tongue:

File: bluekai/config.py [hunk 4]
```diff
@@ -59,11 +59,16 @@ def remote_filename(config):
     partnerName = config.get('BLUEKAI_PARTNERNAME')
     clientName = config.get('BLUEKAI_CLIENTNAME')
     siteId = config.get('BLUEKAI_SITEID')
+    extension = 'bzip2'
```
Comment (ebeesonjanrain) [inline bluekai/config.py]: normally, bzip2 files have the extension '.bz2'

File: bluekai/bluekai_writter.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    def __init__(self, writter, config):
+        (self.data_filename, self.trigger_filename) = remote_filename(config)
+        self.data_fp = writter.file(self.data_filename, mode='w')
+        self.trigger_fp = writter.file(self.trigger_filename, mode='w')
+    def write(self, data):
```
Comment (ebeesonjanrain) [inline bluekai/bluekai_writter.py]: should be blank line between functions at least

File: bluekai/bluekai_writter.py (hunk 2 above)
Comment (ebeesonjanrain) [inline bluekai/bluekai_writter.py]: ok

File: bluekai/config.py [hunk 6]
```diff
@@ -59,11 +59,16 @@ def remote_filename(config):
...
 
     if partnerName and clientName and siteId:
-        return "{}_{}_{}_{}".format(partnerName, clientName, siteId, date)
+        data_filename = "{}_{}_{}_{}.{}".format(partnerName, clientName, siteId, date, extension)
 
     elif partnerName and siteId:
-        return "{}_{}_{}".format(partnerName, siteId, date)
+        data_filename = "{}_{}_{}.{}".format(partnerName, siteId, date, extension)
+
+    trigger_filename = '{}.trigger'.format(data_filename)
```
Comment (BCox-Janrain) [inline bluekai/config.py]: Not super big, but `trigger_filename` will error if partnerName or siteId is missing.

File: bluekai/job.py [hunk 7]
```diff
@@ -38,26 +39,34 @@ def do_job(job, writter, config, logger, datalib, converter):
 
     records_iterator = recordsNewerThan(capture_app, config, last_updated)
 
-    with writter as fp:
+    output = ""
+    record_num = 0
+    for record_num, record in enumerate(records_iterator, start=1):
```
Comment (BCox-Janrain) [inline bluekai/job.py]: What's `record_num = 0` doing? You have it being declared in the for loop right after.

File: bluekai/job.py [hunk 8]
```diff
@@ -38,26 +39,34 @@ def do_job(job, writter, config, logger, datalib, converter):
...
-            except TypeError as exception:
-                message = str(exception)
-                logger.error(message)
-                raise SystemExit(message)
+        output += row
 
-            fp.write(row)
+        if record_num % config['JANRAIN_BATCH_SIZE'] == 0:
+            with closing(writterFactory()) as fp:
+                fp.write(output)
```
Comment (BCox-Janrain) [inline bluekai/job.py]: Does closing context open a new writter for you?

File: bluekai/job.py (hunk 7 above)
Comment (cameron-white-janrain) [inline bluekai/job.py]: I am declaring it out side the for block so that it is still in scope after the for block.

File: bluekai/config.py (hunk 6 above)
Comment (cameron-white-janrain) [inline bluekai/config.py]: As it should

File: bluekai/config.py (hunk 4 above)
Comment (cameron-white-janrain) [inline bluekai/config.py]: Do you want me to change this. Bluekai does not care either way.

Comment (ebeesonjanrain) [inline bluekai/config.py]: If Bluekai doesn't care, I don't either

**Pull request merged**